File: vol3/vol3-java-examples/src/main/java/com/heatonresearch/aifh/randomize/MultiplyWithCarryGenerateRandom.java
/*
 * Artificial Intelligence for Humans
 * Volume 3: Deep Learning and Neural Networks
 * Java Version
 * http://www.aifh.org
 * http://www.jeffheaton.com
 *
 * Code repository:
 * https://github.com/jeffheaton/aifh
 *
 * Copyright 2014-2015 by Jeff Heaton
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * For more information on Heaton Research copyrights, licenses
 * and trademarks visit:
 * http://www.heatonresearch.com/copyright
 */
package com.heatonresearch.aifh.randomize;

/**
 * In Multiply with Carry (MWC) is a pseudo random number generator computer science, multiply-with-carry (MWC)
 * is a method invented by George Marsaglia for generating sequences of random integers based on an initial set
 * from two to many thousands of randomly chosen seed values. The main advantages of the MWC method are that it
 * invokes simple computer integer arithmetic and leads to very fast generation of sequences of random numbers
 * with immense periods.
 * <p/>
 * <p/>
 * This class was implemented using information from the following sources:
 * <p/>
 * http://www.javaprogrammingforums.com/blogs/helloworld922/11-complimentary-multiply-carry-better-way-generate-pseudo-random-numbers.html
 * http://en.wikipedia.org/wiki/Multiply-with-carry
 */
public class MultiplyWithCarryGenerateRandom extends AbstractBoxMuller {
    private long c;
    private long multiplier;
    private int n;
    private int r;
    private final long[] seed;

    public MultiplyWithCarryGenerateRandom(final long seed) {
        this(new long[]{seed}, seed / 2, 64, 987657110L);
    }

    public MultiplyWithCarryGenerateRandom() {
        this(new long[]{System.currentTimeMillis()}, System.nanoTime(), 64, 987657110L);
    }

    public MultiplyWithCarryGenerateRandom(long[] seeds, final long carry, final int r, final long multiplier) {
        setR(r);
        setMultiplier(multiplier);
        this.seed = new long[r];
        if (seeds == null || seeds.length == 0) {
            seeds = new long[]{System.currentTimeMillis()};
        }

        final LinearCongruentialRandom rnd = new LinearCongruentialRandom(seeds[0]);
        this.c = (carry & 0xFFFFFFFFL) % multiplier;
        for (int i = 0; i < r; ++i) {
            if (i < seeds.length) {
                this.seed[i] = seeds[i] & 0xFFFFFFFFL;
            } else {
                this.seed[i] = rnd.nextInt() & 0xFFFFFFFFL;
            }
            if (this.seed[i] == 0xFFFFFFFFL) {
                this.seed[i] = 1L;
            }
        }
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public double nextDouble() {
        return (((long) next(26) << 27) + next(27))
                / (double) (1L << 53);
    }

    private int next(final int bits) {
        final long t = this.multiplier * this.seed[this.n] + this.c;
        final long d32 = t >>> 32;
        this.c = d32 + ((t & 0xFFFFFFFFL) >= 0xFFFFFFFFL - d32 ? 1L : 0L);
        this.seed[this.n] = 0xFFFFFFFEL - (t & 0xFFFFFFFFL) - (this.c - d32 << 32) - this.c & 0xFFFFFFFFL;
        final long result = this.seed[this.n];
        this.n = this.n + 1 & this.r - 1;
        return (int) (result >>> 32 - bits);
    }

    private void setMultiplier(final long theMultiplier) {
        this.multiplier = theMultiplier;
    }

    private void setR(int theR) {
        if (theR <= 0) {
            theR = 256;
        } else {
            boolean validR = true;
            long a = theR;
            while (a != 1 && validR) {
                if (a % 2 != 0) {
                    theR = 256;
                    validR = false;
                }
                a >>>= 1;
            }
        }
        this.r = theR;
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public long nextLong() {
        return ((long) next(32) << 32) + next(32);
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public boolean nextBoolean() {
        return nextDouble() > 0.5;
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public float nextFloat() {
        return (float) nextDouble();
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public int nextInt() {
        return (int) nextLong();
    }
}

File: vol3/vol3-csharp-examples/AIFH-Vol3-Core/Core/Randomize/MultiplyWithCarryGenerateRandom.cs
// Artificial Intelligence for Humans
// Volume 3: Deep Learning and Neural Networks
// C# Version
// http://www.aifh.org
// http://www.jeffheaton.com
//
// Code repository:
// https://github.com/jeffheaton/aifh
//
// Copyright 2015 by Jeff Heaton
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// For more information on Heaton Research copyrights, licenses
// and trademarks visit:
// http://www.heatonresearch.com/copyright
//

using System;

namespace AIFH_Vol3.Core.Randomize
{
    /// <summary>
    ///     Multiply with Carry (MWC) is a pseudo random number generator computer science, multiply-with-carry (MWC)
    ///     is a method invented by George Marsaglia for generating sequences of random integers based on an initial set
    ///     from two to many thousands of randomly chosen seed values. The main advantages of the MWC method are that it
    ///     invokes simple computer integer arithmetic and leads to very fast generation of sequences of random numbers
    ///     with immense periods.
    ///     This class was implemented using information from the following sources:
    ///     http://www.javaprogrammingforums.com/blogs/helloworld922/11-complimentary-multiply-carry-better-way-generate-pseudo-random-numbers.html
    ///     http://en.wikipedia.org/wiki/Multiply-with-carry
    /// </summary>
    public class MultiplyWithCarryGenerateRandom : AbstractBoxMuller
    {
        private readonly long[] _seed;
        private long _c;
        private long _multiplier;
        private int _n;
        private int _r;

        public MultiplyWithCarryGenerateRandom(long seed)
            : this(new[] {seed}, seed/2, 64, 987657110L)
        {
        }

        public MultiplyWithCarryGenerateRandom()
            : this(new long[] {Environment.TickCount}, Environment.TickCount*64, 64, 987657110L)
        {
        }

        public MultiplyWithCarryGenerateRandom(long[] seeds, long carry, int r, long multiplier)
        {
            SetR(r);
            SetMultiplier(multiplier);
            _seed = new long[r];
            if (seeds == null || seeds.Length == 0)
            {
                seeds = new long[] {Environment.TickCount};
            }

            var rnd = new LinearCongruentialRandom(seeds[0]);
            _c = (carry & 0xFFFFFFFFL)%multiplier;
            for (var i = 0; i < r; ++i)
            {
                if (i < seeds.Length)
                {
                    _seed[i] = seeds[i] & 0xFFFFFFFFL;
                }
                else
                {
                    _seed[i] = rnd.NextInt() & 0xFFFFFFFFL;
                }
                if (_seed[i] == 0xFFFFFFFFL)
                {
                    _seed[i] = 1L;
                }
            }
        }

        /**
         * {@inheritDoc}
         */

        public override double NextDouble()
        {
            return (((long) Next(26) << 27) + Next(27))
                   /(double) (1L << 53);
        }

        private int Next(int bits)
        {
            var t = _multiplier*_seed[_n] + _c;
            var d32 = t >> 32;
            _c = d32 + ((t & 0xFFFFFFFFL) >= 0xFFFFFFFFL - d32 ? 1L : 0L);
            _seed[_n] = 0xFFFFFFFEL - (t & 0xFFFFFFFFL) - (_c - d32 << 32) - _c & 0xFFFFFFFFL;
            var result = _seed[_n];
            _n = _n + 1 & _r - 1;
            return (int) (result >> 32 - bits);
        }

        private void SetMultiplier(long theMultiplier)
        {
            _multiplier = theMultiplier;
        }

        private void SetR(int theR)
        {
            if (theR <= 0)
            {
                theR = 256;
            }
            else
            {
                var validR = true;
                long a = theR;
                while (a != 1 && validR)
                {
                    if (a%2 != 0)
                    {
                        theR = 256;
                        validR = false;
                    }
                    a >>= 1;
                }
            }
            _r = theR;
        }

        /**
         * {@inheritDoc}
         */

        public override long NextLong()
        {
            return ((long) Next(32) << 32) + Next(32);
        }

        /**
         * {@inheritDoc}
         */

        public override bool NextBoolean()
        {
            return NextDouble() > 0.5;
        }

        /**
         * {@inheritDoc}
         */

        public override float NextFloat()
        {
            return (float) NextDouble();
        }

        /**
         * {@inheritDoc}
         */

        public override int NextInt()
        {
            return (int) NextLong();
        }
    }
}